File: src/skillguard/governance/policy_engine.py
"""Policy engine for organizational security policies.

Evaluates scan results against configurable organizational policies
to enforce security standards. Supports YAML-based policy definitions
with optional OPA (Open Policy Agent) integration for complex policies.
"""

from __future__ import annotations

from datetime import datetime, timezone
from typing import Any

import yaml
from pydantic import BaseModel, Field

from skillguard.core.models import ScanResult, Severity, Verdict


class PolicyRule(BaseModel):
    """A single policy rule."""

    id: str
    name: str
    description: str = ""
    action: str = "block"  # block, warn, audit
    conditions: dict[str, Any] = Field(default_factory=dict)
    enabled: bool = True


class PolicyViolation(BaseModel):
    """A policy violation found during evaluation."""

    policy_id: str
    policy_name: str
    action: str
    description: str
    details: str = ""


class PolicyResult(BaseModel):
    """Result of policy evaluation."""

    allowed: bool
    violations: list[PolicyViolation] = Field(default_factory=list)
    warnings: list[PolicyViolation] = Field(default_factory=list)
    evaluated_at: str = Field(default_factory=lambda: datetime.now(timezone.utc).isoformat())


class PolicyEngine:
    """Evaluates scan results against organizational policies.

    Supports configurable policies that can block, warn, or audit
    based on scan results. Policies are defined in YAML format.
    """

    def __init__(self) -> None:
        self._policies: list[PolicyRule] = []
        self._load_default_policies()

    def _load_default_policies(self) -> None:
        """Load built-in default policies."""
        self._policies = [
            PolicyRule(
                id="POL-001",
                name="Block Malicious Skills",
                description="Block any skill with a MALICIOUS verdict",
                action="block",
                conditions={"verdict": "malicious"},
            ),
            PolicyRule(
                id="POL-002",
                name="Block Critical Findings",
                description="Block skills with critical severity findings",
                action="block",
                conditions={"min_severity": "critical", "min_count": 1},
            ),
            PolicyRule(
                id="POL-003",
                name="Warn on High Risk",
                description="Warn on skills with HIGH_RISK verdict",
                action="warn",
                conditions={"verdict": "high_risk"},
            ),
            PolicyRule(
                id="POL-004",
                name="Warn on Suspicious",
                description="Warn on skills with SUSPICIOUS verdict",
                action="warn",
                conditions={"verdict": "suspicious"},
            ),
            PolicyRule(
                id="POL-005",
                name="Block High Score",
                description="Block skills with composite score >= 80",
                action="block",
                conditions={"min_score": 80},
            ),
            PolicyRule(
                id="POL-006",
                name="Require OWASP Coverage",
                description="Warn if OWASP LLM01 is detected",
                action="warn",
                conditions={"owasp_detected": "LLM01"},
            ),
        ]

    def load_policies_from_yaml(self, yaml_content: str) -> int:
        """Load policies from a YAML string.

        Returns the number of policies loaded.
        """
        try:
            data = yaml.safe_load(yaml_content)
        except yaml.YAMLError:
            return 0

        if not isinstance(data, dict) or "policies" not in data:
            return 0

        count = 0
        for entry in data.get("policies", []):
            if isinstance(entry, dict) and "id" in entry:
                try:
                    policy = PolicyRule(**entry)
                    self._policies.append(policy)
                    count += 1
                except Exception:
                    continue
        return count

    async def evaluate(self, result: ScanResult) -> PolicyResult:
        """Evaluate a scan result against all active policies.

        Returns:
            PolicyResult with allowed status and any violations.
        """
        violations: list[PolicyViolation] = []
        warnings: list[PolicyViolation] = []

        verdict_str = result.verdict if isinstance(result.verdict, str) else result.verdict.value

        for policy in self._policies:
            if not policy.enabled:
                continue

            violated = self._check_conditions(policy, result, verdict_str)
            if not violated:
                continue

            violation = PolicyViolation(
                policy_id=policy.id,
                policy_name=policy.name,
                action=policy.action,
                description=policy.description,
                details=violated,
            )

            if policy.action == "block":
                violations.append(violation)
            elif policy.action == "warn":
                warnings.append(violation)
            else:  # audit
                warnings.append(violation)

        return PolicyResult(
            allowed=len(violations) == 0,
            violations=violations,
            warnings=warnings,
        )

    def _check_conditions(
        self, policy: PolicyRule, result: ScanResult, verdict_str: str
    ) -> str:
        """Check if a policy's conditions are met. Returns details string if violated."""
        conditions = policy.conditions

        # Check verdict condition
        if "verdict" in conditions:
            if verdict_str == conditions["verdict"]:
                return f"Verdict is {verdict_str}"

        # Check minimum score
        if "min_score" in conditions:
            if result.composite_score >= conditions["min_score"]:
                return f"Score {result.composite_score} >= {conditions['min_score']}"

        # Check minimum severity count
        if "min_severity" in conditions and "min_count" in conditions:
            sev = conditions["min_severity"]
            count = result.findings_by_severity.get(sev, 0)
            if count >= conditions["min_count"]:
                return f"{count} {sev} findings (threshold: {conditions['min_count']})"

        # Check OWASP detection
        if "owasp_detected" in conditions:
            owasp_id = conditions["owasp_detected"]
            if owasp_id in result.owasp_coverage:
                return f"OWASP {owasp_id} detected"

        return ""

    def add_policy(self, policy: PolicyRule) -> None:
        """Add a policy rule."""
        self._policies.append(policy)

    def remove_policy(self, policy_id: str) -> bool:
        """Remove a policy by ID."""
        initial = len(self._policies)
        self._policies = [p for p in self._policies if p.id != policy_id]
        return len(self._policies) < initial

    def list_policies(self) -> list[PolicyRule]:
        """List all configured policies."""
        return list(self._policies)

File: go-skillguard/internal/governance/policy_engine.go
// Package governance provides policy engine, RBAC, and audit logging.
package governance

import (
	"fmt"
	"strings"
	"sync"

	"github.com/girdav01/skillguard/internal/core"
)

// PolicyAction is the action taken when a policy matches.
type PolicyAction string

const (
	PolicyActionBlock PolicyAction = "block"
	PolicyActionWarn  PolicyAction = "warn"
	PolicyActionAudit PolicyAction = "audit"
)

// Policy defines a governance policy.
type Policy struct {
	ID          string       `json:"id" yaml:"id"`
	Name        string       `json:"name" yaml:"name"`
	Description string       `json:"description" yaml:"description"`
	Conditions  []Condition  `json:"conditions" yaml:"conditions"`
	Action      PolicyAction `json:"action" yaml:"action"`
	Enabled     bool         `json:"enabled" yaml:"enabled"`
}

// Condition is a single policy condition.
type Condition struct {
	Field    string `json:"field" yaml:"field"`
	Operator string `json:"operator" yaml:"operator"`
	Value    any    `json:"value" yaml:"value"`
}

// PolicyResult is the result of policy evaluation.
type PolicyResult struct {
	PolicyID string       `json:"policy_id"`
	Name     string       `json:"name"`
	Action   PolicyAction `json:"action"`
	Matched  bool         `json:"matched"`
	Message  string       `json:"message"`
}

// PolicyEngine manages and evaluates governance policies.
type PolicyEngine struct {
	mu       sync.RWMutex
	policies []Policy
}

// NewPolicyEngine creates a new PolicyEngine with default policies.
func NewPolicyEngine() *PolicyEngine {
	pe := &PolicyEngine{}
	pe.loadDefaults()
	return pe
}

func (pe *PolicyEngine) loadDefaults() {
	pe.policies = []Policy{
		{ID: "default-001", Name: "Block Malicious", Description: "Block skills with malicious verdict", Conditions: []Condition{{Field: "verdict", Operator: "eq", Value: "malicious"}}, Action: PolicyActionBlock, Enabled: true},
		{ID: "default-002", Name: "Block High Risk", Description: "Block skills with high risk verdict", Conditions: []Condition{{Field: "verdict", Operator: "eq", Value: "high_risk"}}, Action: PolicyActionBlock, Enabled: true},
		{ID: "default-003", Name: "Warn Suspicious", Description: "Warn on suspicious skills", Conditions: []Condition{{Field: "verdict", Operator: "eq", Value: "suspicious"}}, Action: PolicyActionWarn, Enabled: true},
		{ID: "default-004", Name: "Block Critical Findings", Description: "Block skills with critical findings", Conditions: []Condition{{Field: "critical_count", Operator: "gt", Value: 0}}, Action: PolicyActionBlock, Enabled: true},
		{ID: "default-005", Name: "Warn High Score", Description: "Warn on score above 50", Conditions: []Condition{{Field: "score", Operator: "gt", Value: 50}}, Action: PolicyActionWarn, Enabled: true},
		{ID: "default-006", Name: "Audit All Scans", Description: "Audit log all scan results", Conditions: []Condition{{Field: "score", Operator: "gte", Value: 0}}, Action: PolicyActionAudit, Enabled: true},
	}
}

// ListPolicies returns all policies.
func (pe *PolicyEngine) ListPolicies() []Policy {
	pe.mu.RLock()
	defer pe.mu.RUnlock()
	result := make([]Policy, len(pe.policies))
	copy(result, pe.policies)
	return result
}

// AddPolicy adds a new policy.
func (pe *PolicyEngine) AddPolicy(p Policy) {
	pe.mu.Lock()
	defer pe.mu.Unlock()
	pe.policies = append(pe.policies, p)
}

// RemovePolicy removes a policy by ID.
func (pe *PolicyEngine) RemovePolicy(id string) bool {
	pe.mu.Lock()
	defer pe.mu.Unlock()
	for i, p := range pe.policies {
		if p.ID == id {
			pe.policies = append(pe.policies[:i], pe.policies[i+1:]...)
			return true
		}
	}
	return false
}

// Evaluate evaluates all policies against a scan result.
func (pe *PolicyEngine) Evaluate(result *core.ScanResult) []PolicyResult {
	pe.mu.RLock()
	defer pe.mu.RUnlock()

	var results []PolicyResult
	for _, policy := range pe.policies {
		if !policy.Enabled {
			continue
		}
		matched := pe.evaluateConditions(policy.Conditions, result)
		msg := ""
		if matched {
			msg = fmt.Sprintf("Policy '%s' triggered: %s", policy.Name, policy.Description)
		}
		results = append(results, PolicyResult{
			PolicyID: policy.ID,
			Name:     policy.Name,
			Action:   policy.Action,
			Matched:  matched,
			Message:  msg,
		})
	}
	return results
}

func (pe *PolicyEngine) evaluateConditions(conditions []Condition, result *core.ScanResult) bool {
	for _, cond := range conditions {
		if !pe.evaluateCondition(cond, result) {
			return false
		}
	}
	return true
}

func (pe *PolicyEngine) evaluateCondition(cond Condition, result *core.ScanResult) bool {
	switch cond.Field {
	case "verdict":
		actual := string(result.Verdict)
		expected, _ := cond.Value.(string)
		return compareString(actual, cond.Operator, expected)
	case "score":
		return compareInt(result.CompositeScore, cond.Operator, toInt(cond.Value))
	case "critical_count":
		count := result.FindingsBySeverity["critical"]
		return compareInt(count, cond.Operator, toInt(cond.Value))
	case "high_count":
		count := result.FindingsBySeverity["high"]
		return compareInt(count, cond.Operator, toInt(cond.Value))
	case "total_findings":
		return compareInt(result.TotalFindings, cond.Operator, toInt(cond.Value))
	}
	return false
}

func compareString(actual, op, expected string) bool {
	switch op {
	case "eq":
		return strings.EqualFold(actual, expected)
	case "neq":
		return !strings.EqualFold(actual, expected)
	case "contains":
		return strings.Contains(strings.ToLower(actual), strings.ToLower(expected))
	}
	return false
}

func compareInt(actual int, op string, expected int) bool {
	switch op {
	case "eq":
		return actual == expected
	case "neq":
		return actual != expected
	case "gt":
		return actual > expected
	case "gte":
		return actual >= expected
	case "lt":
		return actual < expected
	case "lte":
		return actual <= expected
	}
	return false
}

func toInt(v any) int {
	switch val := v.(type) {
	case int:
		return val
	case int64:
		return int(val)
	case float64:
		return int(val)
	case string:
		n := 0
		fmt.Sscanf(val, "%d", &n)
		return n
	}
	return 0
}
